ton_lib/src/tep/meta/meta_loader.rs

```diff
@@ -1 +1,163 @@
+        self.http_client = Some(http_client);
+        self
+    }
+    pub fn with_ipfs_loader(mut self, ipfs_loader: IpfsLoader) -> Self {
+        self.ipfs_loader = Some(ipfs_loader);
+        self
+    }
+
+    pub fn build(self) -> MetaLoader {
+        let mut headers = header::HeaderMap::new();
+        headers.insert("user-agent", HeaderValue::from_static("TonlibMetaLoader/0.x"));
+        headers.insert("accept", HeaderValue::from_static("*/*"));
+
+        MetaLoader {
+            http_client: self.http_client.or(Client::builder().default_headers(headers).build().ok()).unwrap(),
+            ipfs_loader: self.ipfs_loader.or(IpfsLoader::new(&IpfsLoaderConfig::default()).ok()).unwrap(),
+        }
+    }
+}
+
+impl MetaLoader {
+    pub fn builder() -> MetaLoaderBuilder {
+        MetaLoaderBuilder::new()
+    }
+
+    pub async fn load_json_meta_from_uri(&self, uri: &str) -> Result<String, MetaLoaderError> {
+        log::trace!("Downloading metadata from {}", uri);
+        let meta_str: String = if uri.starts_with("ipfs://") {
+            let path: String = uri.chars().skip(7).collect();
+            self.ipfs_loader.load_utf8_lossy(path.as_str()).await?
+        } else {
+            let resp = self.http_client.get(uri).send().await?;
+            if resp.status().is_success() {
+                resp.text().await?
+            } else {
+                return Err(MetaLoaderError::LoadMetadataFailed {
+                    uri: uri.to_string(),
+                    status: resp.status(),
+                });
+            }
+        };
+
+        Ok(meta_str)
+    }
+}
+
+// ------------------------------------JETTON META DATA----------------------------------------
+impl MetaLoader {
+    pub async fn load<T: Metadata>(&self, content: &MetadataContent) -> Result<T, MetaLoaderError> {
+        match content {
+            MetadataContent::External(MetadataExternal { uri }) => {
+                let json = self.load_json_meta_from_uri(&uri.as_str()).await?;
+                Ok(T::from_offchain(&json)?)
+            }
+            MetadataContent::Internal(MetadataInternal { data: dict }) => {
+                if dict.contains_key(&*META_URI) {
+                    let uri = String::from_utf8_lossy(dict.get(&*META_URI).unwrap().as_slice()).to_string();
+                    match self.load_json_meta_from_uri(uri.as_str()).await {
+                        Ok(json) => Ok(T::from_data(Some(&dict), Some(&json))?),
+                        Err(_) => Ok(T::from_onchain(&dict)?),
+                    }
+                } else {
+                    Ok(T::from_onchain(&dict)?)
+                }
+            }
+            content => Err(MetaLoaderError::ContentLayoutUnsupported(content.clone())),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use anyhow::Ok;
+
+    use crate::tep::meta::meta_loader::{MetaLoader, NftItemMetadata};
+
+    #[tokio::test]
+    async fn test_meta_data_load_ordinal_https() -> anyhow::Result<()> {
```

Comment:
It's integration tests (with external service) and should be put into `tests` folder